### PR TITLE
Track model interactions

### PR DESCRIPTION
**Pull request by lucagobbi**
# Description

This PR aims to prepare the groundwork for tracking model interactions and resource usage. 

@pieroit I believe this feature is a valuable addition to the framework because, especially for newcomers, the Cat is often treated as a black box, logs can sometimes be challenging to interpret. Providing a comprehensive observability of what is really happening under the hood can signficantly improve the dev ex with the framework.

However, Im still facing some ancestral doubts:

- Should `ModelInteraction` be a unique class? Since the embedder model interaction type has no output tokens, a fixed source (`recall`, is it the only one?) and a fixed type for reply `List[float]`, maybe it should be better to split `ModelInteraction` into two separated classes `LLMModelInteraction` and `EmbedderModelInteraction`?
- Another central issue is the standardized tokenizer. Following @pieroit 
suggestions in the issue,  I used `tiktoken`, the OpenAI tokenizer, for both input tokens and output tokens (to avoid excluding non-streaming APIs). I've stumbled upon different discussions online and it seems that the community is facing an error of `±15%` depending on the provider (here a [link](https://github.com/gkamradt/LLMTest_NeedleInAHaystack/issues/25) to an interesting discussion about that). I feel like we're taking a very rigid approach by using tiktoken everywhere, I've experimented also with response metadata but Langchain is still generic about [that](https://api.python.langchain.com/en/latest/outputs/langchain_core.outputs.llm_result.LLMResult.html#langchain_core.outputs.llm_result.LLMResult) and providers are not standardizing their responses.

Tested with OpenAI (gpt-4o, gpt-3.5-turbo), Cohere (command), Ollama (gemma2:9b).

Related to issue #864 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: core/cat/looking_glass/callbacks.py [hunk 1]
```diff
@@ -8,3 +11,41 @@ def __init__(self, stray):
...
+                source=source,
+                prompt="",
+                reply="",
+                input_tokens=0,
+                output_tokens=0,
+            )
+        )
+
+    def _count_tokens(self, text: str) -> int:
+        # cl100k_base is the most common encoding for OpenAI models such as GPT-3.5, GPT-4 - what about other providers?
```
Comment (pieroit) [inline core/cat/looking_glass/callbacks.py:35]: Awesome, let's stick with this for the moment!

**Comment by pieroit** [comment]
Hi @lucagobbi thanks for this, awesome feat

> @pieroit I believe this feature is a valuable addition to the framework because, especially for newcomers, the Cat is often treated as a black box, logs can sometimes be challenging to interpret. Providing a comprehensive observability of what is really happening under the hood can signficantly improve the dev ex with the framework.

Agree, also noobs will require less support from us

> * Should `ModelInteraction` be a unique class? Since the embedder model interaction type has no output tokens, a fixed source (`recall`, is it the only one?) and a fixed type for reply `List[float]`, maybe it should be better to split `ModelInteraction` into two separated classes `LLMModelInteraction` and `EmbedderModelInteraction`?

Your call!
I guess it can be a good idea. What we care about is an easy interpretation of what happened in the backend, so the focus is on devx.
Another pro of separating is to have a `report` including both calls to models and calls to other parts of the framework (hooks, tools, vector memory - which can be many). Can we generalize? A list of interactions, all subclassing a main class.
I suggest in the main class we also add a timestamp

> * Another central issue is the standardized tokenizer. Following @pieroit
>   suggestions in the issue,  I used `tiktoken`, the OpenAI tokenizer, for both input tokens and output tokens (to avoid excluding non-streaming APIs). I've stumbled upon different discussions online and it seems that the community is facing an error of `±15%` depending on the provider (here a [link](https://github.com/gkamradt/LLMTest_NeedleInAHaystack/issues/25) to an interesting discussion about that). I feel like we're taking a very rigid approach by using tiktoken everywhere, I've experimented also with response metadata but Langchain is still generic about [that](https://api.python.langchain.com/en/latest/outputs/langchain_core.outputs.llm_result.LLMResult.html#langchain_core.outputs.llm_result.LLMResult) and providers are not standardizing their responses.

@lucagobbi @valentimarco @zAlweNy26 
On the admin report we can show `~3200 tokens` to indicate that it is an estimation.
Most people needs this to make calculations on vendor spend for a typical interaction, so I would not worry much about the estimate being exact, at least until langchain itself offers the information

**Comment by lucagobbi** [comment]
I've updated the PR splitting the `ModelInteraction` class into:

- `LLMModelInteraction` class 
- `EmbedderModelInteraction` class

Base `ModelInteraction` class has now a `started_at` timestamp, while LLMModelInteraction has also an `ended_at` timestamp to measure response duration (is it useful?)

@pieroit sorry I dont actually get what you mean by `report`, is this moving in the right direction?

**Comment by pieroit** [comment]
> I've updated the PR splitting the `ModelInteraction` class into:
> 
> * `LLMModelInteraction` class
> * `EmbedderModelInteraction` class
> 
> Base `ModelInteraction` class has now a `started_at` timestamp, while LLMModelInteraction has also an `ended_at` timestamp to measure response duration (is it useful?)

Awesome thanks @lucagobbi !

> 
> @pieroit sorry I dont actually get what you mean by `report`, is this moving in the right direction?

Yes, I mean instead of only model interactions, we log in the list also other stuff (tools, hooks, vector memory queries).
This can be the next version of the `why`.
I can merge this for the moment and we later discuss how to include other info.

**Comment by pieroit** [comment]
So sweet 

Screenshot from 2024-07-16 12-45-39

**Comment by zAlweNy26** [comment]
I'll upgrade the Admin panel to show this :)

**Comment by pieroit** [comment]
> I'll upgrade the Admin panel to show this :)

Propose a wireframe bedore coding please

**Pull request merged**